s3-plugin-server/src/main/scala/com/gu/teamcity/S3.scala

```diff
@@ -29,15 +26,41 @@ class S3(config: S3ConfigManager) {
     (for (bucket <- targetBucket) yield
       Try {
         val uploadDirectory = s"${S3Plugin.cleanFullName(build)}/${build.getBuildNumber}"
-        val metadata = {
-          val md = new ObjectMetadata()
-          md.setContentLength(fileSize)
-          md
-        }
-        val req = new PutObjectRequest(bucket, s"$uploadDirectory/$fileName", contents, metadata)
+        val req: PutObjectRequest = putRequestAsFile(bucket, s"$uploadDirectory/$fileName", contents)
         val upload = transferManager.upload(req)
         upload.waitForUploadResult()
         true
       }
-    ) getOrElse (Success(false))
+    ) getOrElse Success(false)
+
+  def putRequestAsFile(bucket: String, targetName: String, contents: InputStream): PutObjectRequest = {
+    // convert to a file, then the S3 client can do parallel upload, and also not fail with reset stream problems
+    val file = MakeFile(contents)
+    val req = new PutObjectRequest(bucket, targetName, file)
+    contents.close()
+    req
+  }
+
 }
+
+/**
+ * Return the input stream as a File, consuming everything but not closing it.
```

Comment:
it's not really my stream to close.  But then again it's also not really my stream to read things from.  But closing just seemed a bit presumptuous.  Ahh the life in a mutable world...